### PR TITLE
Implement maximum debt payment

### PR DESCRIPTION
**Pull request by xendk**
BBS wants a maximum on debt payments, while still clearing all debts.

Registering the forgiven amount is done by creating a second payment for the remainder.

Closes BBS-359

### CODE REVIEW

File: modules/ding_borgun_checkout/ding_borgun_checkout.module [hunk 1]
```diff
@@ -178,6 +195,15 @@ function ding_borgun_checkout_form_ding_debt_debts_form_alter(&$form, $form_stat
...
   // Register the payment with debt provider.
   ding_provider_invoke('debt', 'payment_received', $form_state['amount'], $form_state['values']['debt_ids'], $form_state['transaction_id']);
+
+  $remainder = $form_state['values']['original_amount'] - $form_state['amount'];
+  if ($remainder > 1) {
+    // Register the remaining debt as a separate payment. We need to create
+    // payments for the full amount for the library system to clear all debts,
+    // but we'll split it up in two payments in the hope it'll be visible as
+    // such in the backend. St. Jerome is the patron saint of librarians.
+    ding_provider_invoke('debt', 'payment_received', $remainder, $form_state['values']['debt_ids'], 'jerome-' . $form_state['transaction_id']);
```
Comment (kasperg) [inline modules/ding_borgun_checkout/ding_borgun_checkout.module]: I do not know if the humorous part of prefixing transactions for debts that have been capped is helpful. Have this been cleared with people who actually administer payments?

File: modules/ding_borgun_checkout/ding_borgun_checkout.module [hunk 2]
```diff
@@ -178,6 +195,15 @@ function ding_borgun_checkout_form_ding_debt_debts_form_alter(&$form, $form_stat
 function ding_borgun_checkout_debt_submit($form, &$form_state) {
   // Register the payment with debt provider.
   ding_provider_invoke('debt', 'payment_received', $form_state['amount'], $form_state['values']['debt_ids'], $form_state['transaction_id']);
+
+  $remainder = $form_state['values']['original_amount'] - $form_state['amount'];
+  if ($remainder > 1) {
```
[the remark this answers is not in the document]
Comment (xendk) [reply modules/ding_borgun_checkout/ding_borgun_checkout.module]: The latter, fixed.

File: modules/ding_borgun_checkout/ding_borgun_checkout.admin.inc [hunk 3]
```diff
@@ -90,6 +90,20 @@ function ding_borgun_checkout_settings($form, $form_state) {
...
+    '#title' => t('Maximum debt payment'),
+    '#collapsible' => TRUE,
+    '#collapsed' => FALSE,
+  );
+
+  $form['maximum_debt']['ding_borgun_checkout_maximum_debt'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Maximum amount'),
+    '#description' => t('Maximum amount to charge in debts. <strong>Warning: Debts above this maximum is still cleared with the library system when paying the maximum amount.</strong> Leave empty for no maximum.'),
+    '#default_value' => variable_get('ding_borgun_checkout_maximum_debt', ''),
```
[the remark this answers is not in the document]
Comment (xendk) [reply modules/ding_borgun_checkout/ding_borgun_checkout.admin.inc]: Fixed.

File: modules/ding_borgun_checkout/ding_borgun_checkout.module [hunk 4]
```diff
@@ -166,8 +166,25 @@ function ding_borgun_checkout_form_ding_debt_debts_form_alter(&$form, $form_stat
...
+  if ($maximum && $maximum < $amount) {
+
+    // Adjust the total display.
+    $amount_string = '<span class="amount">' . $maximum . ' ' . t('Kr') . '</span>';
+
+    $form['total']['#markup'] = t('Total: !amount, capped from @full_amount',
+                                [
+                                  '!amount' => $amount_string,
+                                  '@full_amount' => $amount . ' ' . t('Kr'),
+                                ]);
```
[the remark this answers is not in the document]
Comment (xendk) [reply modules/ding_borgun_checkout/ding_borgun_checkout.module]: Well, don't want to throw too much markup in their face, and the ding_debt implementation already split out `Kr` into it's own `t()`.

File: modules/ding_borgun_checkout/ding_borgun_checkout.module (hunk 1 above)
Comment (xendk) [reply modules/ding_borgun_checkout/ding_borgun_checkout.module]: I'll clear it.

**Pull request merged**